daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceChildParseResultHelper.scala

```diff
@@ -0,0 +1,477 @@
+    isZL: Boolean,
+    erd: ElementRuntimeData,
+    requiredOptional: RequiredOptionalStatus): ParseAttemptStatus =
+    anyTypeElementFailedParseAttemptStatus(pstate, isZL, requiredOptional)
+}
+
+trait ModelGroupSequenceChildParseResultHelper
+  extends SequenceChildParseResultHelper {
+
+  def mgrd: ModelGroupRuntimeData
+
+  /**
+   * The model group's representation could be zero length.
+   *
+   * False if the model group is not represented (contains no syntax, and
+   * all content is recursively not represented)
+   *
+   * True if the model group has no mandatory syntax, and all content within
+   * is possibly ZL (recursively), defaultable, optional, or not represented.
+   */
+  def isModelGroupRepPossiblyZeroLength: Boolean
+
+  /**
+   * The model group's representation must be greater than zero length.
+   */
+  def isModelGroupRepNonZeroLength: Boolean
+
+  /**
+   * a group is always required.
+   */
+  final override def maybeStaticRequiredOptionalStatus: Maybe[RequiredOptionalStatus] =
+    Maybe(RequiredOptionalStatus.Required)
+
+  /**
+   * Compute the ParseAttemptStatus, given the state of the parse immediately after parsing
+   * the item (which could be group or element).
+   */
+  final override def computeParseAttemptStatus(parser: SequenceChildParser,
+    prevBitPosBeforeChild: Long,
+    pstate: PState,
+    requiredOptional: RequiredOptionalStatus): ParseAttemptStatus = {
+    val currentBitPosAfterChild = pstate.bitPos0b
+    val isZL = {
+      Assert.invariant(currentBitPosAfterChild >= prevBitPosBeforeChild)
+      currentBitPosAfterChild == prevBitPosBeforeChild
+    }
+    if (pstate.isSuccess) {
+      checkModelGroupZL(pstate, isZL)
+      modelGroupSuccessParseAttemptStatus(parser, pstate, isZL, mgrd, requiredOptional)
+    } else {
+      Assert.invariant(pstate.isFailure)
+      computeFailedParseAttemptStatus(parser, prevBitPosBeforeChild, pstate, isZL, requiredOptional)
+    } // end if isSuccess/isFailed
+  }
+
+  /**
+   * Did the most recent parse succeed consuming zero length, and is that allowed?
+   */
+  final protected def checkModelGroupZL(pstate: PState, isZL: Boolean): Unit = {
+    Assert.invariant(pstate.isSuccess)
+    val isIt = isZL
+    if (isZL) {
+      // This doesn't hold if
+      // (a) the format has a terminator which is an expression
+      // (b) the expression evaluates to say, %WSP*; or %ES; based on looking at other infoset information.
+      // (c) that delimiter matches zero length
+      // (d) the lengthKind is NOT delimited. So we're not scanning for this.
+      // This comes up in mil-std-2045 and other formats which have an optional
+      // final terminator after a string having lengthKind 'pattern'.
+      // In that case, the static information would indicate positively that isModelGroupRepNonZeroLength
+      // is true, when it isn't.
+      //
+      // Assert.invariant(isModelGroupRepPossiblyZeroLength && !isModelGroupRepNonZeroLength)
```

Comment:
So is the issue here that our static logic isn't yet complete, and so we can't detect some of these edge cases?